lib/internal/http2/compat.js

```diff
@@ -290,17 +297,30 @@ class Http2ServerResponse extends Stream {
     stream.on('finish', onfinish);
   }
 
+  // User land modules such as finalhandler just check truthiness of this
+  // but if someone is actually trying to use this for more than that
+  // then we simply can't support such use cases
+  get _header() {
+    return this.headersSent;
+  }
+
   get finished() {
     const stream = this[kStream];
-    return stream === undefined || stream._writableState.ended;
+    return stream === undefined ||
+           stream._writableState.ended ||
+           this[kState].closed;
   }
 
-  get closed() {
-    return this[kState].closed;
+
+  get socket() {
```

Comment:
I'm wondering if we shouldn't more proactively warn people away from accessing this... e.g. by using an explicit runtime deprecation warning right from the start.

Reply:
Similar to statusMessageWarn? We could definitely do that.

Reply:
yep, exactly like that.

Reply:
As long as there is a way to retrieve the remote IP address for the socket (and local port, etc.); ideally symmetrically with http/1.

Reply:
I think we will keep the socket available but just add a one-time warning re: its usage.

Reply:
Good point and a good reminder! I had that marked down quite some time ago as a todo for an API that would make fetching those details easier but hadn't gotten around to it. We can make that data available via the `Http2Session` object so that the `socket` property does not have to be accessed.

Reply:
> I think we will keep the socket available but just add a one-time warning re: its usage.

So getting the remote IP will print a warning?

Reply:
*just* getting it via the request and response objects should print a warning. We should add apis that allow this information to be retrieved independently of that so that there's a path that (a) does not print the warning and (b) does not mean users are accessing the socket directly.

Reply:
Sounds good. Would that land prior to landing a warning on the `socket` getter, then?

Reply:
We could probably do those at the same time actually.